test/UniV3LoanRouter.ts

```diff
@@ -391,7 +392,492 @@ describe("Uniswap V3 Loan Router", () => {
+
+    const trancheImplementation = <Tranche>await deploy("Tranche", admin, []);
+    const trancheFactory = <TrancheFactory>await deploy("TrancheFactory", admin, [trancheImplementation.address]);
+
+    const bondImplementation = <BondController>await deploy("BondController", admin, []);
+    const bondFactory = <BondFactory>(
+      await deploy("BondFactory", admin, [bondImplementation.address, trancheFactory.address])
+    );
+
+    const tranches = [200, 300, 500];
+    let bond: BondController | undefined;
+    const tx = await bondFactory
+      .connect(admin)
+      .createBond(mockWrapperToken.address, tranches, await time.secondsFromNow(10000));
+    const receipt = await tx.wait();
+    if (receipt && receipt.events) {
+      for (const event of receipt.events) {
+        if (event.args && event.args.newBondAddress) {
+          bond = <BondController>await hre.ethers.getContractAt("BondController", event.args.newBondAddress);
+        }
+      }
+    } else {
+      throw new Error("Unable to create new bond");
+    }
+    if (!bond) {
+      throw new Error("Unable to create new bond");
+    }
+
+    const trancheContracts: Tranche[] = [];
+    for (let i = 0; i < tranches.length; i++) {
+      const tranche = <Tranche>await hre.ethers.getContractAt("Tranche", (await bond.tranches(i)).token);
+      trancheContracts.push(tranche);
+    }
+    // load up the mock uniswap with tokens for swapping
+    await mockCollateralToken.mint(mockSwapRouter.address, hre.ethers.utils.parseEther("1000000000000"));
+    await mockCashToken.mint(mockSwapRouter.address, hre.ethers.utils.parseEther("1000000000000"));
+
+    return {
+      router,
+      mockCollateralToken,
+      mockWrapperToken,
+      bond,
+      bondFactory,
+      tranches: trancheContracts,
+      mockCashToken,
+      user,
+      other,
+      signers: signers.slice(2),
+    };
+  };
+
+  describe("wrapAndBorrowMax", function () {
+    it("should successfully wrap and borrow max", async () => {
+      const { router, tranches, mockCollateralToken, mockWrapperToken, mockCashToken, bond, user } = await loadFixture(
+        fixture,
+      );
+      const amount = hre.ethers.utils.parseEther("100");
+      await mockCollateralToken.connect(user).mint(await user.getAddress(), amount);
+      await mockCollateralToken.connect(user).approve(router.address, amount);
+
+      // min output of 50 because the router will sell all A (20) and B (30) tranche tokens, but keep the Z tranches
+      const minOutput = hre.ethers.utils.parseEther("50");
+      await expect(
+        router
+          .connect(user)
+          .wrapAndBorrowMax(amount, bond.address, mockCashToken.address, minOutput, { gasLimit: 9500000 }),
+      )
+        .to.emit(mockCollateralToken, "Transfer")
+        .withArgs(await user.getAddress(), router.address, amount)
+        .to.emit(mockWrapperToken, "Transfer")
+        .withArgs(router.address, bond.address, amount)
```

Comment:
I think it'd help to add comments that explain why certain assumptions can be made based on the mock behaviour.
For example here - the mock wrapper gives the user as much wrapped tokens as the tokens they deposited. This isn't a guarantee (or likely) with the actual wrappers, so reasoning about this test requires understanding that mock behaviour.

Reply:
(Similarly this test relies on A and B tranche tokens being sold 1:1 with the loan currency, which threw me for a loop until I reasoned it through in the context of mocks)

Reply:
added a couple comments - good point!